### ISSUE TITLE
running a bash with rstudio user and specified USERID

### ISSUE THREAD

**Issue by kforner** [rocker-org/rocker-versioned2#332]
Probably a trivial question, but I could not find an answer in the README nor on the https://www.rocker-project.org site.

How to run a bash with a non-root user (e.g. _rstudio_) and a given USERID (e.g. matching the host user) in the container ?

I noticed that if I run rstudio like this:
```
docker run --rm -ti -p 8787:8787 -e USERID=$(id -u) -e DISABLE_AUTH=true  rocker/verse:4.1.2
```
if works perfectly and the _rstudio_ user, for example in the rstudio terminal, has the right **userid**.

But if I execute: 
```
docker run --rm -ti -e USERID=$(id -u) -u rstudio rocker/verse:4.1.2 bash
```
the _rstudio_ user id has not changed, it is still 1000. 
How could I have this user id fixed ?

**Comment by eddelbuettel** [comment]
We do that in the PrairieLearn use case where we have both user ids 1000 and 1001 and pivot to 1001. I can probably dig that up when I have a moment, or we ask @echuber2 who can probably recite it from memory.  Maybe as 'simple' as ensureing 1001 actually exists.

**Comment by echuber2** [comment]
As Dirk mentioned, for PrairieLearn's use case we did something more hacky. Since the `init_userconf.sh` will delete and recreate the account if you set `USERID`, instead, we edited the old 1000 account to be 1001 inherently:

https://github.com/PrairieLearn/PrairieLearn/blob/8007f528f5d976d8e11f3952566bca24ccbb9c71/workspaces/rstudio/Dockerfile#L9-L17

We intentionally avoided setting `USERID` then (noted here):

https://github.com/PrairieLearn/PrairieLearn/blob/8007f528f5d976d8e11f3952566bca24ccbb9c71/workspaces/rstudio/Dockerfile#L94-L97

My reason for doing this was that our modified image, or images derived from it, has (or might have) stuff manually placed in the home directory that shouldn't be wiped out or recreated at startup time.

**Comment by cboettig** [comment]
Thanks Eric, those are good points!  arguably that is more reasonable for `init_userconf.sh` too -- happy to look at a PR if you're interested!

**Comment by kforner** [comment]
For what it's worth, I was using this perl script to do a similar task. Note that the usermod will automatically update the files in the user's home directory

```perl
#!/usr/bin/env perl
use strict;
use warnings;

# script to change the user and group id of the docker user so that it matches
# the ids of the user running the docker to help sharing files between the host
# and the guest

my ($USER, $HOST_UID, $HOST_GID, $GROUP) = @ARGV;

$GROUP ||= 'qbstaff';

sub usage {
  my $msg = shift || "";
  $msg .= "\n" if $msg;
  my $usage =
  qq{USAGE: $0 username uid gid [groupname]
  };

  die($msg . $usage . "\n");
}

# a stderred friendly log function
sub logme {
  print STDERR "@_\n";
}

usage("Give gid") unless($HOST_GID);

my $uid = getpwnam $USER;
my $gid = getgrnam $USER;

if ($uid ne $HOST_UID) {
  logme("fixing USER $USER uid from $uid to $HOST_UID");

  # we may need to create the group
  if ($uid ne $HOST_UID) {
    logme("creating group $GROUP with id $HOST_GID");
    my $cmd = qq{addgroup --gid $HOST_GID $GROUP};
    system($cmd) and warn "unable to execute command '$cmd': $!";
  }

  my $cmd = qq{usermod -u $HOST_UID -g $HOST_GID $USER};
  system($cmd) and die "unable to execute command '$cmd': $!";
}



```

**Comment by kforner** [comment]
@cboettig 
> You should be able to merely run the same userconf.sh script manually after the bash session starts,

This script must be run as root. How should I switch to the fixed user ? 
In the rstudio server case, this the server itself that does that for the sessions. 

If I use the `-user` option, the script will not be run as root. 

Currently, what I can foresee is:

add an entrypoint which
   - if some args are given to docker run
      - delete `/etc/services.d/rstudio` to disable the rstudio service
      - execute `userconf.sh`
      - run the args using `su rstudio $ARGS`
   - execute `/init`

What do you think ?

**Comment by kforner** [comment]
thanks @eitsupi 

> I wonder if the --user option of the docker run command can satisfy the need.

I think a problem with that approach is that the user home is empty, so that you can not customize it, with a .bashrc for example, or other user config files. 
Moreover I guess the user does not really have a name which can be a problem in some cases


> modified userconf.sh

adding `su`, and `exec` may probably (not tested) not play well with stdin/stdout streams/

An alternative way, that I used successfully before, is to use perl's builtin feature of changing the user and effective user id of the current process, here´s an example:

```perl
sub switch_to_user {
  my ($userid, $groupid) = @_;
	logme("switching to user $userid and groupid $groupid");

	### set group id (setgid)
	# N.B must be before the setuid
	$( = $groupid;
	die "setgid failed: $!" unless $( == $groupid;

	# set the effective uid
	$> = $userid;
	if ($> != $userid) {die "seteuid failed: $!\n"}

	# set the real uid
	$< = $userid;
	if ($< != $userid) {die "setuid failed: $!\n"}
}
```

This way you can switch the user id of the current process, reusing its open stdin/stdout/stderr.

**Comment by eitsupi** [comment]
> I think a problem with that approach is that the user home is empty, so that you can not customize it, with a .bashrc for example, or other user config files.

When doing such customization, I believe it is common for users to create their own Dockerfile, and it should be easy to create a new UID 1001 user in the Dockerfile.

Why do you want to run bash with user UID 1001?
For example, if the difference in UID between the user on the Docker host and the user in the container (in this case `rstudio`) due to bind mounting is a problem, we can use docker volumes without bind mounting (this is docker's recommendation).
https://docs.docker.com/storage/volumes/

**Comment by kforner** [comment]
@eitsupi 
> Why do you want to run bash with user UID 1001?

I want the user to be able to edit: 
-  its sources files, on its disk,  with rstudio or another IDE/editor running in the image
- to run code in the container that generates output files on the disk.
-
I don't want UID 1001 but to match the host user id.
Imagine that I have scores of users and several docker images based on rocker. 
This has to happen automatically. 
My users almost don't even have to know there are using docker behind the scene  ...

**Comment by eitsupi** [comment]
How about attaching to a container after you have created it?

```shell
$ docker run --rm -d -e USERID=1001 --name test rocker/rstudio
35e8a4e9c90160d55bce79a116af7e576d0ed7bfeebbf8652e642262f57b8d1b
$ docker exec -it -u rstudio test bash
rstudio@35e8a4e9c901:/$ id
uid=1001(rstudio) gid=1001(rstudio) groups=1001(rstudio),50(staff)
rstudio@35e8a4e9c901:/$ exit
exit
$ docker stop test
test
```

**Comment by kforner** [comment]
@eitsupi Thanks for the suggestion. 
That's interesting but that won't do: 
- I would have to manage the identification shutdown of the daemon container
- that would run the rstudio server for nothing, consuming resources

That's really nice of you, but I thought it was a common use case and I was probably wrong. 
Most people will probably just use rstudio. 
I will use a custom endpoint like I did previously. 
FYI, I got information from the developer of S6, that this use case is currently not supported but will be in the next version.

Again, thank you all for your help.
